File: conversions/merge-deep/elide-merge-deep.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Ruby Integration Example for elide-merge-deep
#
# This demonstrates calling the TypeScript merge-deep implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One merge implementation shared across TypeScript and Ruby
# - Consistent deep merging across services
# - No Ruby gem needed
# - Guaranteed format consistency

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# merge_module = Elide.require('./elide-merge-deep.ts')

puts "=== Ruby Calling TypeScript Merge Deep ==="
puts

# Example 1: Rails Configuration
# default_config = {
#   database: {
#     adapter: 'postgresql',
#     host: 'localhost',
#     port: 5432
#   },
#   cache: {
#     store: :memory_store
#   }
# }
#
# env_config = {
#   database: {
#     host: 'db.example.com',
#     database: 'production'
#   }
# }
#
# merged = merge_module.call(:default, default_config, env_config)
# puts "Merged Rails config: #{merged}"
# puts

# Example 2: Sidekiq Job Configuration
# class DataProcessorJob
#   include Sidekiq::Job
#
#   def perform(user_id, options = {})
#     default_options = {
#       retry: true,
#       timeout: 300,
#       callbacks: {
#         on_success: :log_success,
#         on_error: :log_error
#       }
#     }
#
#     # Merge with user options using Elide
#     final_options = merge_module.call(:default, default_options, options)
#
#     process_data(user_id, final_options)
#   end
# end

# Example 3: Deep Hash Merging
# user_settings = {
#   profile: { name: 'Alice', age: 25 },
#   preferences: { theme: 'dark', language: 'en' }
# }
#
# updates = {
#   preferences: { theme: 'light' }
# }
#
# merged_settings = merge_module.call(:default, user_settings, updates)
# puts "Merged settings: #{merged_settings}"
# puts

# Example 4: Array Merge Strategies
# data1 = { tags: ['ruby', 'rails'] }
# data2 = { tags: ['elide', 'polyglot'] }
#
# # Replace strategy (default)
# replaced = merge_module.call(:default, data1, data2)
# puts "Replaced: #{replaced}"
#
# # Concat strategy
# options = { arrayMerge: 'concat' }
# concatenated = merge_module.call(:mergeDeepWith, options, data1, data2)
# puts "Concatenated: #{concatenated}"
#
# # Unique strategy
# options = { arrayMerge: 'unique' }
# unique = merge_module.call(:mergeDeepWith, options, data1, data2)
# puts "Unique: #{unique}"

# Example 5: API Client Configuration
# class ApiClient
#   DEFAULT_CONFIG = {
#     base_url: 'https://api.example.com',
#     timeout: 30,
#     headers: {
#       'Content-Type' => 'application/json'
#     },
#     retry: {
#       max_attempts: 3,
#       backoff: :exponential
#     }
#   }.freeze
#
#   def initialize(custom_config = {})
#     @config = merge_module.call(:default, DEFAULT_CONFIG, custom_config)
#   end
# end

puts "Real-world use case:"
puts "- Ruby Rails/Sidekiq apps merge configs"
puts "- Uses same TypeScript implementation as Node.js service"
puts "- Guarantees consistent config format across entire stack"
puts "- No need to install Ruby deep-merge gem"
puts

puts "Example: E-commerce Platform"
puts "┌─────────────────────────────────────┐"
puts "│   Elide Merge-Deep (TypeScript)    │"
puts "│   elide-merge-deep.ts               │"
puts "└─────────────────────────────────────┘"
puts "         ↓                   ↓"
puts "    ┌────────┐          ┌────────┐"
puts "    │ Node.js│          │  Ruby  │"
puts "    │  API   │          │Workers │"
puts "    └────────┘          └────────┘"
puts "         ↓                   ↓"
puts "    Same deep merge logic everywhere!"
puts

puts "Problem Solved:"
puts "Before: Different merge implementations = subtle bugs"
puts "After: One Elide implementation = 100% consistent"
puts

puts "Performance Benefits:"
puts "- Zero cold start overhead with Elide"
puts "- Instant deep merging"
puts "- Shared runtime across languages"
puts "- No Ruby gem installation needed"
puts

puts "When Elide Ruby API is ready, usage will be:"
puts "  merge_module = Elide.require('./elide-merge-deep.ts')"
puts "  result = merge_module.call(:default, obj1, obj2)  # That's it!"

File: conversions/merge-deep/elide-merge-deep.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-merge-deep

This demonstrates calling the TypeScript merge-deep implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One merge implementation shared across TypeScript and Python
- Consistent deep merging across services
- No Python merge library needed
- Guaranteed format consistency
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# merge_module = require('./elide-merge-deep.ts')

print("=== Python Calling TypeScript Merge Deep ===\n")

# Example 1: Configuration Merging
# default_config = {
#     'server': {'port': 3000, 'host': 'localhost'},
#     'database': {'name': 'mydb'}
# }
#
# user_config = {
#     'server': {'port': 4000},
#     'database': {'user': 'admin', 'pass': 'secret'}
# }
#
# merged = merge_module.default(default_config, user_config)
# print(f"Merged config: {merged}")
# print()

# Example 2: Deep State Update
# current_state = {
#     'user': {
#         'profile': {'name': 'Alice', 'age': 25},
#         'preferences': {'theme': 'dark', 'notifications': True}
#     }
# }
#
# state_update = {
#     'user': {
#         'preferences': {'theme': 'light'}
#     }
# }
#
# new_state = merge_module.default(current_state, state_update)
# print(f"New state: {new_state}")
# print()

# Example 3: Flask/Django Configuration
# from flask import Flask
#
# app = Flask(__name__)
#
# # Base configuration
# base_config = {
#     'DEBUG': False,
#     'TESTING': False,
#     'DATABASE': {
#         'ENGINE': 'postgresql',
#         'HOST': 'localhost',
#         'PORT': 5432
#     }
# }
#
# # Environment-specific configuration
# env_config = {
#     'DEBUG': True,
#     'DATABASE': {
#         'HOST': 'db.example.com',
#         'NAME': 'production_db'
#     }
# }
#
# # Merge configurations using Elide
# app.config.update(merge_module.default(base_config, env_config))

# Example 4: Data Pipeline Configuration
# def load_pipeline_config(base_config, overrides):
#     """Load and merge pipeline configuration"""
#     merged = merge_module.default(base_config, overrides)
#
#     print(f"Pipeline config loaded:")
#     print(f"  Source: {merged['source']}")
#     print(f"  Destination: {merged['destination']}")
#     print(f"  Transforms: {len(merged['transforms'])} steps")
#
#     return merged

# Example 5: Array Merge Strategies
# data1 = {'items': ['a', 'b']}
# data2 = {'items': ['c', 'd']}
#
# # Replace strategy (default)
# replaced = merge_module.default(data1, data2)
# print(f"Replaced: {replaced}")
#
# # Concat strategy
# concatenated = merge_module.mergeDeepWith({'arrayMerge': 'concat'}, data1, data2)
# print(f"Concatenated: {concatenated}")
#
# # Unique strategy
# unique = merge_module.mergeDeepWith({'arrayMerge': 'unique'}, data1, data2)
# print(f"Unique: {unique}")

print("Real-world use case:")
print("- Python Flask/Django apps merge configs")
print("- Uses same TypeScript implementation as Node.js service")
print("- Guarantees consistent config format across entire stack")
print("- No need to install Python deep-merge library")
print()

print("Example: Microservices Architecture")
print("┌─────────────────────────────────────┐")
print("│   Elide Merge-Deep (TypeScript)    │")
print("│   elide-merge-deep.ts               │")
print("└─────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │ Worker │")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same deep merge logic everywhere!")
print()

print("Problem Solved:")
print("Before: Different merge implementations = inconsistent behavior")
print("After: One Elide implementation = 100% consistent merging")
print()

print("Performance Benefits:")
print("- Zero cold start overhead with Elide")
print("- Instant deep merging")
print("- Shared runtime across languages")
print("- No Python library installation needed")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  merge = require('./elide-merge-deep.ts')")
print("  result = merge.default(obj1, obj2)  # That's it!")
